src/cpp/_pythia8.cpp

```diff
@@ -0,0 +1,225 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <pybind11/iostream.h>
+#include <Pythia8/Event.h>
+#include <Pythia8/Pythia.h>
+#include <Pythia8/ParticleData.h>
+#include <Pythia8/Info.h>
+#include <Pythia8/HIUserHooks.h>
+#include <array>
+#include <cassert>
+
+namespace py = pybind11;
+using namespace Pythia8;
+using namespace pybind11::literals;
+
+struct Hepevt
+{
+    Hepevt() { reset(10000); }
+
+    void reset(int size)
+    {
+        idhep = py::array_t<int>({size});
+        isthep = py::array_t<int>({size});
+        phep = py::array_t<double>({5, size});
+        vhep = py::array_t<double>({4, size});
+        jmohep = py::array_t<int>({2, size});
+        jdahep = py::array_t<int>({2, size});
+    }
+
+    void fill(const Event &event)
+    {
+        ++nevhep;
+        nhep = event.size();
+        if (nhep > py::len(idhep))
+            reset(nhep);
+
+        auto pid = idhep.mutable_unchecked<1>();
+        auto sta = isthep.mutable_unchecked<1>();
+        auto p = phep.mutable_unchecked<2>();
+        auto v = vhep.mutable_unchecked<2>();
+        auto par = jmohep.mutable_unchecked<2>();
+        auto dau = jdahep.mutable_unchecked<2>();
+        int k = 0;
+        for (const auto &particle : event)
+        {
+            // Record starts with internal particle 90,
+            // skip this to start with beam particles.
+            // This results in Fortran-like indices.
+            if (k == 0 && particle.id() == 90)
+            {
+                --nhep;
+                continue;
+            }
+            pid(k) = particle.id();
+            sta(k) = particle.statusHepMC();
+            p(0, k) = particle.px();
+            p(1, k) = particle.py();
+            p(2, k) = particle.pz();
+            p(3, k) = particle.e();
+            p(4, k) = particle.m();
+            v(0, k) = particle.xProd();
+            v(1, k) = particle.yProd();
+            v(2, k) = particle.zProd();
+            v(3, k) = particle.tProd();
+            par(0, k) = particle.mother1();
+            par(1, k) = particle.mother2();
+            dau(0, k) = particle.daughter1();
+            dau(1, k) = particle.daughter2();
+            ++k;
+        }
+    }
+
+    int nevhep = -1;
+    int nhep = 0;
+    py::array_t<int, py::array::f_style> idhep;
+    py::array_t<int, py::array::f_style> isthep;
+    py::array_t<double, py::array::f_style> phep;
+    py::array_t<double, py::array::f_style> vhep;
+    py::array_t<int, py::array::f_style> jmohep;
+    py::array_t<int, py::array::f_style> jdahep;
+};
+
+double charge_from_pid(const ParticleData &pd, int pid)
+{
+    auto pptr = pd.findParticle(pid);
+    assert(pptr); // never nullptr if charge_from_pid is used on particles produced by Pythia
+    // ParticleData returns partice even if anti-particle pid is used
+    return pid == pptr->id() ? pptr->charge() : -pptr->charge();
```

Comment:
I rather would like to get rid of the `charge` attribute in EventData and make that a property which is computed dynamically from the pid, using a fast function like this. This would fix the SIBYLL-2.1 test issue that we have currently and simplify the implementation of MCEvent.

Reply:
ok